Initial commit of mindstorms_widgets class and demo script

dandrus-ev3 opened the pull request:
Denis, I'm interested in your feedback on this.  I'm still interested in working to improve the documentation to make python-ev3dev more approachable for people who don't know C++ or much about linux sysfs attributes, but it occurred to me that a mindstorms_widgets class would have the benefit of leveraging the Lego documentation.  This is the first time I've contributed to a non-work related project so I'm very open to any feedback you have about my code.  If you like it, this may be a good addition to your demo folder...

dandrus-ev3 committed:
Initial commit of mindstorms_widgets class and demo script

ddemidov commented:
Hi Don!

I like the idea behind `mindstorms_widget` class. I initially went for something similar with [ev3dev_utils.motors](https://github.com/ddemidov/ev3dev-lang-python/blob/0.0.4/ev3dev_utils/motors.py) module (now deleted), but not to the extent you did it here.

I found eventually that using the unified API directly is more straightforward/convenient, but I agree the class may be useful for people coming from the original ev3 software (and may be even as a backend for a graphical programming system, e.g. ev3dev/ev3dev#348). I think we could include it into python-ev3dev  package as a submodule, so that people would not have to clone the repo to use it.

> a mindstorms_widgets class would have the benefit of leveraging the Lego documentation

Can you access the documentation online? I only saw it from inside the original lego software, and that is very inconvenient, at least for me, since I have to reboot to windows to use it.

ddemidov commented on demo/mindstorms_widgets.py:
```diff
@@ -0,0 +1,321 @@
+
+    # Begin public methods
+        
+    def connect_motor( self, motor ):
```
Why not add a third parameter here, `port=None`, and stick to the default behavior when its not provided? This way we could allow non-standard connections as well.

ddemidov commented on demo/mindstorms_widgets.py:
```diff
@@ -0,0 +1,321 @@
+            #led.red_left.flash( 200 )
+            #todo:Seems that the delay_on/delay_off attributes aren't there in
+            #/sys/class/leds/ev3-left0:red:ev3dev/
+            pass
```
You need to use the timer trigger to see the attributes, but even after that `flash()` seems to be broken. I've opened an issue for this: ev3dev/ev3dev-lang#92. For now, the following should work:

``` py
led.red_left.trigger = 'timer'
led.red_left.set_on_delay(200)
led.red_left.set_off_delay(200)
```

ddemidov replied:
`flash()` should be fixed in the latest version (see ev3dev/ev3dev-lang#92). Forget my comment about setting the trigger to `timer` manually, since flash does that anyway.

ddemidov commented on demo/mindstorms_widgets.py:
```diff
@@ -0,0 +1,321 @@
+            
+        for m,p in zip ( motors, powers ):
+            # Negative power and invert are synonomous, so the easiest
+            # logic is to invert the polarity when negative power is requested
```
I am not sure about this. First, using invert instead of negative power will only work if users will stick to this function for all their interactions with the motor. Second, this is simply more work: you need to set two attributes (polarity and power) instead of one (power).

dandrus-ev3 replied:
So you are saying it is surprising to the user if they put a negative power but I convert it to a positive power and toggle the invert attribute for them... I can see that.  This trick avoided some other complexity related to using the steering function but I can look into fixing that a different way.

ddemidov commented:
I've made a couple of inline comments.

ddemidov commented on demo/mindstorms_widgets_demo.py:
```diff
@@ -0,0 +1,255 @@
+# display even though brickman is still running.
+
+# Need advice on best way to avoid constantly being prompted for sudo password
+os.system("sudo chvt 7")
```
If you run the script from the brickman menu (the `File system` one), then `chvt` call should not be necessary. But you can avoid a password prompt by changing `/etc/sudoers` file.

dandrus-ev3 commented:
Denis, thanks for your input.  Hopefully I'll have some more time this weekend to make updates.  I can't find the Mindstorms documentation online either.  It seems to only be visible through the app.

dandrus-ev3 committed:
Added optional port parameter to connect_motor()

Also tried to use LED flash() method but still ran into issues, and
tried to avoid ugly side effect of reversing motor polarity rather than
setting sysfs attributes intuitively, but eventually gave up and just
documented the limitation.

dandrus-ev3 commented:
Hi Denis.  I tried to use the updated LED flash().  It worked the fist time I called it but subsequent calls caused a crash.  I also tried to avoid the unintuitive way I'm inverting the motor polarity but the solution was so ugly I backed it out.  I documented why I think my solution is the least of evils, and am hoping that anyone using mindstorms_widgets directly will not need to simultaneously control the motors directly...  Initially I was hoping this would serve as complicated demo for others like me who would otherwise find the unified API hard to figure out, and I think it accomplishes that.  I'm also willing to do more work on it but at the rate I'm going it will be October before I would be able to upload another round of changes.

BTW, I'm coaching a robotics team for my 4th grade son's elementary and our First Lego League (FLL) competition starts next week.  They will be using mindstorms to complete a set of missions defined by Lego, and compete with other schools.  I  will be using your interface to code the missions in Python, to show them how it is done without giving them any code they could steal (and without having to use the nifty but extremely frustrating mindstorms software...

ddemidov commented:
Thanks, Dan, I merged the PR.

> BTW, I'm coaching a robotics team for my 4th grade son's elementary and our First Lego League (FLL) competition starts next week. 

Good luck!

ddemidov commented:
I'll try to experiment with the led flash later.

ddemidov commented:
Dan, the issue with led::flash should be solved by ev3dev/ev3dev-lang@7dc3aff25b027e90730937dd442d376208961444.
